src/protocol/boolean/carries.rs

```diff
@@ -0,0 +1,405 @@
+    ///
+    /// # Example
+    /// ```ignore
+    ///   // Input
+    ///   [a] = 1 1 0 1 0 1 1 0   // 214 in LE
+    ///   [b] = 0 0 0 1 1 1 0 0   // 28 in LE
+    ///   // Output
+    ///   [s] = 0 0 0 1 0 1 0 0
+    /// ```
+    async fn step1<F: Field>(
+        a: &[Replicated<F>],
+        b: &[Replicated<F>],
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let mul = zip(repeat(ctx), zip(a, b))
+            .enumerate()
+            .map(|(i, (ctx, (&a_bit, &b_bit)))| {
+                let c = ctx.narrow(&BitOpStep::Step(i));
+                async move { c.multiply(record_id, a_bit, b_bit).await }
+            });
+        try_join_all(mul).await
+    }
+
+    /// Step 2. for `i=0..l=1`, generate a vector of tuples `([s], [p], [k])` where
+    /// `[p_i] = [a_i] + [b_i] - 2*[s_i]`, and `[k_i] = 1 - [s_i] - [p_i]`
+    ///
+    /// The interpretations is:
+    ///   * `s_i = 1` iff carry is set at position `i` (i.e. `a_i + b_i = 2`)
+    ///   * `p_i = 1` iff carry would be propagated at position `i` (i.e. `a_i + b_i = 1`)
+    ///   * `k_i = 1` iff a carry would be killed at position `i` (i.e. `a_i + b_i = 0`)
+    ///
+    /// Therefore, we compute `p_i = XOR(a_i, b_i)`, and `k_i = !OR(a_i, b_i)`
+    ///
+    /// # Example
+    /// ```ignore
+    /// // Input
+    ///   [a] = 1 1 0 1 0 1 1 0   // 214 in LE
+    ///   [b] = 0 0 0 1 1 1 0 0   // 28 in LE
+    ///   [s] = 0 0 0 1 0 1 0 0
+    /// // Output
+    ///   [s] = 0 0 0 1 0 1 0 0   // `a_i & b_i`
+    ///   [p] = 1 1 0 0 1 0 1 0   // `a_i ^ b_i`
+    ///   [k] = 0 0 1 0 0 0 0 1   // `!(a_i | b_i)`
+    /// ```
+    #[allow(clippy::many_single_char_names)]
+    fn step2<F: Field>(
+        a: &[Replicated<F>],
+        b: &[Replicated<F>],
+        s: &[Replicated<F>],
+        one: Replicated<F>,
+    ) -> Vec<CarryPropagationShares<F>> {
+        zip(a, b)
+            .zip(s)
+            .map(|((&a_bit, &b_bit), &s_bit)| {
+                let p_bit = a_bit + b_bit - (s_bit * F::from(2));
+                CarryPropagationShares {
+                    s: s_bit,
+                    p: p_bit,
+                    k: one - s_bit - p_bit,
+                }
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// Step 3. Prefix Carry-Propagation
+    ///
+    /// Let `e_i = (s_i, p_i, k_i)`, and `○` be the carry-propagation operator, it
+    /// computes `([f_0],...,[f_l]) ← PRE○([e_0],...,[e_l])`.
+    async fn step3<F: Field>(
+        e: &[CarryPropagationShares<F>],
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<Vec<CarryPropagationShares<F>>, BoxError> {
+        let l = e.len();
+        let futures = (0..l).map(|i| {
+            let c = ctx.narrow(&BitOpStep::Step(i));
+            async move { Self::fan_in_carry_propagation(&e[0..=i], c, record_id).await }
```

Comment:
I am 100% confident that this is doing a ton of duplicate work. I'm OK landing it as-is, but we should attempt to optimize it later.